File: bindings/javascript/src/safe_queue.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeQueue - Bounded FIFO queue.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 * @module
 */

import { getLib, ProvenStatus, statusToError } from './ffi.js';
import { ok, err } from './result.js';

/**
 * Bounded FIFO queue backed by the libproven FFI.
 */
export class BoundedQueue {
  /** @type {Deno.PointerObject|null} */
  #ptr = null;

  /**
   * Create a bounded queue with the given capacity.
   *
   * @param {number} capacity - Maximum number of elements.
   * @returns {{ ok: true, value: BoundedQueue } | { ok: false, error: string }}
   */
  static create(capacity) {
    const symbols = getLib();
    const ptr = symbols.proven_queue_create(capacity);
    if (ptr === null) return err('Failed to create queue');
    const queue = new BoundedQueue();
    queue.#ptr = ptr;
    return ok(queue);
  }

  /**
   * Push a value to the queue.
   *
   * @param {number|bigint} value - The value to push.
   * @returns {boolean} True if push succeeded (queue not full).
   */
  push(value) {
    if (this.#ptr === null) return false;
    const symbols = getLib();
    return symbols.proven_queue_push(this.#ptr, BigInt(value));
  }

  /**
   * Pop a value from the queue.
   *
   * @returns {{ ok: true, value: number } | { ok: false, error: string }}
   */
  pop() {
    if (this.#ptr === null) return err('Queue is closed');
    const symbols = getLib();
    const result = symbols.proven_queue_pop(this.#ptr);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(Number(result[1]));
  }

  /**
   * Get the current size of the queue.
   *
   * @returns {number}
   */
  size() {
    if (this.#ptr === null) return 0;
    const symbols = getLib();
    return Number(symbols.proven_queue_size(this.#ptr));
  }

  /**
   * Close and free the native queue.
   */
  close() {
    if (this.#ptr !== null) {
      const symbols = getLib();
      symbols.proven_queue_free(this.#ptr);
      this.#ptr = null;
    }
  }
}

/**
 * Priority queue (placeholder -- the FFI layer provides FIFO only;
 * priority queue will be added when the Idris implementation is complete).
 */
export class PriorityQueue extends BoundedQueue {}

File: bindings/rust/src/safe_queue.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe bounded queue operations via libproven FFI.
//!
//! Provides FIFO queues with bounded capacity.
//! All operations delegate to Idris 2 verified code.

use crate::core::{self, Error, Result};
use crate::ffi;

/// A bounded FIFO queue.
///
/// The underlying memory is managed by libproven. The queue is freed
/// when this struct is dropped.
pub struct BoundedQueue {
    ptr: *mut ffi::BoundedQueue,
}

// SAFETY: The underlying queue is managed by libproven.
unsafe impl Send for BoundedQueue {}

impl BoundedQueue {
    /// Create a new bounded queue with the given capacity.
    pub fn new(capacity: usize) -> Result<Self> {
        // SAFETY: proven_queue_create takes a value-type usize;
        // always safe to call.
        let ptr = unsafe { ffi::proven_queue_create(capacity) };
        if ptr.is_null() {
            return Err(Error::AllocationFailed);
        }
        Ok(BoundedQueue { ptr })
    }

    /// Push a value to the back of the queue.
    ///
    /// Returns `false` if the queue is full.
    pub fn push(&mut self, value: i64) -> bool {
        // SAFETY: self.ptr is valid (checked at construction).
        unsafe { ffi::proven_queue_push(self.ptr, value) }
    }

    /// Pop a value from the front of the queue.
    ///
    /// Returns `Err` if the queue is empty.
    pub fn pop(&mut self) -> Result<i64> {
        // SAFETY: self.ptr is valid (checked at construction).
        let result = unsafe { ffi::proven_queue_pop(self.ptr) };
        core::int_result_to_result(result)
    }

    /// Get the current number of elements in the queue.
    pub fn size(&self) -> usize {
        // SAFETY: self.ptr is valid (checked at construction).
        unsafe { ffi::proven_queue_size(self.ptr) }
    }

    /// Check if the queue is empty.
    pub fn is_empty(&self) -> bool {
        self.size() == 0
    }
}

impl Drop for BoundedQueue {
    fn drop(&mut self) {
        // SAFETY: self.ptr was allocated by proven_queue_create
        // and has not been freed yet.
        unsafe {
            ffi::proven_queue_free(self.ptr);
        }
    }
}

/// A priority queue (wraps bounded queue with priority semantics).
///
/// Note: The underlying FFI uses the same queue primitives. Priority
/// ordering is handled at the Idris 2 level.
pub struct PriorityQueue {
    inner: BoundedQueue,
}

impl PriorityQueue {
    /// Create a new priority queue with the given capacity.
    pub fn new(capacity: usize) -> Result<Self> {
        Ok(PriorityQueue {
            inner: BoundedQueue::new(capacity)?,
        })
    }

    /// Insert a value (priority encoded in the value).
    pub fn insert(&mut self, value: i64) -> bool {
        self.inner.push(value)
    }

    /// Remove and return the highest-priority value.
    pub fn remove(&mut self) -> Result<i64> {
        self.inner.pop()
    }

    /// Get the current size.
    pub fn size(&self) -> usize {
        self.inner.size()
    }

    /// Check if empty.
    pub fn is_empty(&self) -> bool {
        self.inner.is_empty()
    }
}
